### PR TITLE
feat: add support for asynchronous filters in Nunjucks

### PR DESCRIPTION
**Pull request by magicmatatjahu**
**Description**

Add support for asynchronous filters in Nunjucks:
- check before saving filter to registry that function is asynchronous, if yes, add as third argument to `addFilter` function `true` - it indicates that to filter should be always added callback with signature `callback(err, result)`.
- update docs with example.

In Nunjucks docs we can read:
image

**Related issue(s)**
Resolves https://github.com/asyncapi/generator/issues/515

### CODE REVIEW

**Comment by magicmatatjahu** [comment]
@derberg I have one code smell with line `const asyncFunctionCtor = (async () => {}).constructor;` Unfornatelly this is needed to check that function is async.

**Comment by derberg** [comment]
@magicmatatjahu is there no other way? what about `const isAsync = myFunction.constructor.name === "AsyncFunction";`

**Comment by magicmatatjahu** [comment]
@derberg Comparing constructor's name is usually not a good idea, but I'll see if the sonar has a problem with that.

**Comment by magicmatatjahu** [comment]
@derberg I changed it and it helped to remove the smell code, but I still think that comparing name is not good idea, but I can accept it here.

File: docs/authoring.md [hunk 1]
```diff
@@ -343,4 +343,21 @@ You can use the filter function in your template as in the following example:
 const {{ channelName | camelCase }} = '{{ channelName }}';
 ```
 
+The generator supports also asynchronous filters. Asynchronous filters receive as last argument a callback to resume rendering. Asynchronous filters must be annotated with the `async` keyword. Make sure to call the callback with two arguments: `callback(err, res)`. `err` can be `null`. See the example of how to use asynchronous filters:
```
Comment (jonaslagoni) [inline docs/authoring.md:346]:
```suggestion
The generator also supports asynchronous filters. Asynchronous filters receive as last argument a callback to resume rendering. Asynchronous filters must be annotated with the `async` keyword. Make sure to call the callback with two arguments: `callback(err, res)`. `err` can be `null`. See the following example of how to use asynchronous filters:
```

What about filters that return a promise? 🤔

File: lib/utils.js [hunk 2]
```diff
@@ -150,3 +150,13 @@ utils.getInvalidOptions = (generatorOptions, options) => {
   if (typeof options !== 'object') return [];
   return Object.keys(options).filter(param => !generatorOptions.includes(param));
 };
+
+/**
+ * Determine whether the given `fn` is an async function.
```
Comment (jonaslagoni) [inline lib/utils.js:155]:
```suggestion
 * Determine whether the given function is asynchronous.
```

File: lib/utils.js [hunk 3]
```diff
@@ -150,3 +150,13 @@ utils.getInvalidOptions = (generatorOptions, options) => {
...
+ * @private
+ * @param {*} fn
+ * @returns {Boolean}
```
Comment (jonaslagoni) [inline lib/utils.js:158]:
```suggestion
 * @param {*} fn to check 
 * @returns {Boolean} is function asynchronous
```

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
Just got some small suggestions 😄

File: docs/authoring.md (hunk 1 above)
Comment (magicmatatjahu) [reply docs/authoring.md]: Nunjucks doesn't support it. Determine that function returns Promise is impossible without invoke it, so we must think about another way to resolve problem.

Comment (jonaslagoni) [reply docs/authoring.md]: Yea, let's not focus on that at all, if people want it they will ask for it 😄

**Review by jonaslagoni — DISMISSED** [review]
LGTM 👍

File: docs/authoring.md (hunk 1 above)
Comment (magicmatatjahu) [reply docs/authoring.md]: My first thought how to resolve that: we should return as filter not function but object with metadata that function is asynchronous, like

```js
const filters = module.export;

filters.someFilter = {
  isAsync: true,
  fn: function() { return new Promise() ... }
}
```

File: docs/authoring.md [hunk 4]
```diff
@@ -343,4 +343,21 @@ You can use the filter function in your template as in the following example:
 const {{ channelName | camelCase }} = '{{ channelName }}';
 ```
 
+The generator also supports asynchronous filters. Asynchronous filters receive as last argument a callback to resume rendering. Asynchronous filters must be annotated with the `async` keyword. Make sure to call the callback with two arguments: `callback(err, res)`. `err` can be `null`. See the following example of how to use asynchronous filters:
+
```
Comment (jonaslagoni) [inline docs/authoring.md:347]: Maybe we should add the fact that promise cannot be used 🤔

Comment (magicmatatjahu) [reply docs/authoring.md:347]: Or make this workaround https://github.com/asyncapi/generator/pull/516#discussion_r585493590 for Promises 😄

Comment (jonaslagoni) [reply docs/authoring.md:347]: I feel like it is more a bandage and complicates the filters more than we should.

**Comment by magicmatatjahu** [comment]
@jonaslagoni I updated docs with Promise example :)

**Review by jonaslagoni — APPROVED** [review]

**Pull request merged**